### ISSUE TITLE
Run Out Of Memory on RTX 3090

### ISSUE THREAD

**Issue by YuliangXiaoYLX** [voxelmorph/voxelmorph#426]
**Task** (what are you trying to do/register?)

I wanted to train my own dataset to do the scan-to-scan registration. (GPU: RTX 3090, 24GB memory)

**What have you tried**

I used ```voxelmorph/scripts/train.py``` to obtain the scan-to-scan deformation field model

**Details of experiments**

The CT scans I used are in high resolution which are 512x512x(# of slices). If I did not downsample the scans and start training directly, the GPU memory will run out of. However, if I used ```Scipy``` to pad and downsample images to 256x256x128, the training result was very bad. Can anyone help me solve this issue and provide me some suggestions? Thanks!!!

**Comment by adalca** [comment]
Can you describe in more detail your 256x256x128 runs? 
- How did you normalize data? what loss did you use? how much data did you use to train? etc.

**Comment by YuliangXiaoYLX** [comment]
> Can you describe in more detail your 256x256x128 runs?
> 
> * How did you normalize data? what loss did you use? how much data did you use to train? etc.

@adalca Thanks for your response. 
1. I checked the dimension of each image. As I observed, most of my data has number of slices less than 256 and others are same, which means the third value of its dimension is less than 256 and first two are same. So, if the third value of dimension is less than 256, I will pad it to 256, otherwise, I will crop it to 256. 
2. I used ```Scipy.zoom.interpolation``` to downsample by 2 to get 256x256x128.
3. I normalized CT scan using the 0.05, 95 clipping % z-score normalization. Here is the simple demo:
```
mean = np.mean(scan)
std = np.std(scan)
lb = np.percentile(scan, 0.05)
ub = np.percentile(scan, 99.5)
scan = np.clip(scan, lb, ub)
scan = (scan - mean) / std
```
5. The loss I used was MSE and my training data is 28 scans.

I am looking forward to your response!

**Comment by adalca** [comment]
you should probably take the mean and std after the clip, but otherwise that looks fine (ish). 
What was the loss hyperparameter balancing loss and grad?

**Comment by YuliangXiaoYLX** [comment]
> you should probably take the mean and std after the clip, but otherwise that looks fine (ish). What was the loss hyperparameter balancing loss and grad?

I think I used default parameters of ```train.py``` file except changing the ```grad-loss-weight``` to 0.001

**Comment by adalca** [comment]
ok, and what does it mean that the results are very bad?

**Comment by YuliangXiaoYLX** [comment]
> ok, and what does it mean that the results are very bad?

![image](https://user-images.githubusercontent.com/70731556/167974578-aed388e8-512c-4532-afed-e39654df1c4e.png)
![image](https://user-images.githubusercontent.com/70731556/167974600-9c361ebf-26fa-4d29-8772-e208b7a8af17.png)
The moved image is the registered result, you can clearly see that there are many difference between fixed and moved images

**Comment by YuliangXiaoYLX** [comment]
> you should probably take the mean and std after the clip, but otherwise that looks fine (ish). What was the loss hyperparameter balancing loss and grad?

Btw here is a detailed info of the hyperparameter:

![Screen Shot 2022-05-11 at 21 41 51](https://user-images.githubusercontent.com/70731556/167975009-b25daa50-a284-4241-b6fd-1d75f70956f2.png)

**Comment by adalca** [comment]
I see, some training curves might help. Is that a training subject?

**Comment by YuliangXiaoYLX** [comment]
> I see, some training curves might help. Is that a training subject?

yes, do you have any suggestions to improve the result?

**Comment by adalca** [comment]
can you please provide the training curves

**Comment by YuliangXiaoYLX** [comment]
> 

@adalca Here is the training curves
![image](https://user-images.githubusercontent.com/70731556/168199843-2814067a-3e65-4c44-aba6-d237420dfdf4.png)

**Comment by adalca** [comment]
I would not say that that is converged, quite yet. I would
- run it much longer
- compute the MSE and grad on the image pair you showed me -- is loss around 0.02 as the graph shows? If not, you might be doing something wrong in your visualization

**Comment by ha5750** [comment]
Hi, here's the updated result trained with lambda = 0.001. 
<img width="1645" alt="image" src="https://user-images.githubusercontent.com/74314232/168384939-be653f25-1b38-4593-b9ba-a3dba6b5686d.png">
<img width="1604" alt="image" src="https://user-images.githubusercontent.com/74314232/168385068-8b333ce0-12d9-4ffb-acd6-927613741cec.png">

Question 1) As you can see, some small details become lost in moved images after registration due to high complexity of training images. Do you think there's a room for improvement for registration with VoxelMorph?

Also, our ultimate goal is to segment (not register) the ROI by doing
1) create atlas and manually segment the ROI
2) register the new scan with the atlas
3) warp the atlas ROI to new scan to obtain the segmentation result.

Question 2) Would this be a reasonable way of performing a high-quality segmentation, or would directly learning the ROI (as in nn-UNET) be a better way? 

Question 3) Our dataset generates a sub-par ("fussy" structures) template result for unconditaional template training. Would this mean that the dataset have too many variations?

**Comment by adalca** [comment]
Those results look pretty good to me, although you might want to increase the grad loss a bit. The goal isn't to get the images to perfectly match detail! People are different after al.

If you only have time to segment an anatomical ROI on one subject, Honestly any of those 3 methods will work roughly the same in my opinion. 1 and 2 might provide more regular results with fewer island-type errors. If you  do option 3 you'll need a ton of augmentation like we do in https://github.com/xamyzhao/brainstorm/

**Comment by codeenthusiast** [comment]
Is there any way for us to improve the registration for the finer structures. As you can see the sinuses are definitely not warped as accurately and that is the level of detail we are concerned with. We are able to achieve quite good accuracy with ANTs when doing scan to scan registration but I'm curious as to if we can obtain similar results in VoxelMorph?

**Comment by adalca** [comment]
You should be able to. Can you show the exact same slice and example with ANTs?
